File: 662-maximum-width-of-binary-tree/maximum-width-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Pair{
    TreeNode node;
    int width;
    Pair(TreeNode node, int width){
        this.node = node;
        this.width = width;
    }
}

class Solution {
    public int widthOfBinaryTree(TreeNode root) {
        Queue<Pair> q = new LinkedList<>();
        q.add(new Pair(root,0));
        int maxWidth = 0;

        while(!q.isEmpty()){
            int size = q.size();
            int last = 0 ,first = 0;
            int startIndex = 0;
            for(int i=0;i<size;i++){
                TreeNode cur = q.peek().node;
                int width = q.peek().width;
                q.poll();

                if(i==0){
                    first = width;
                    startIndex = 2*width+1;
                }

                if(i==size-1){
                   last = width;
                }

                if(cur.left!=null)
                    q.add(new Pair(cur.left,2*width+1-startIndex));
                if(cur.right!=null)
                    q.add(new Pair(cur.right,2*width+2-startIndex));
            }
                maxWidth = Math.max(maxWidth,last-first+1); 
        }

        return maxWidth;
    }
}
/*
    0
    1           2
    2  3        4 5
 */

File: 662-maximum-width-of-binary-tree/maximum-width-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def widthOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        max_width = 0
        queue = deque([(root, 0)])
        while queue:
            level_size = len(queue)
            # Set the 'leftmost_index' to the index of the first node in this level. The
            # 'rightmost_index' starts at the same point but gets updated as we iterate
            # over the level, and ends at the index of the last node at this level.
            leftmost_index = queue[0][1]
            rightmost_index = leftmost_index
            start_index = 2*leftmost_index+1
	        # take a new variable start_index to calculate 
            # the index of first node of next level
            # Process all nodes at the current level.
            for _ in range(level_size):
                node, i = queue.popleft()
	     
                if node.left:
                    queue.append((node.left, 2*i + 1-start_index))
                if node.right:
                    queue.append((node.right, 2*i + 2-start_index))
                rightmost_index = i
            max_width = max(max_width, rightmost_index - leftmost_index + 1)
        return max_width
